### PR TITLE
Make it possible to share maps between both directions

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Generalize the conversion functionality to make it possible to use "generic maps" (i.e. `vector<tuple<K, V>>` or a proper `map<K, V>`).
  - This makes most of the conversion a template (i.e. header) library.
  - Keep the current behavior by specifying suitable defaults for all the templates.
- This is necessary to support the introduction of a shared global map in [key4hep/k4MarlinWrapper#147](https://github.com/key4hep/k4MarlinWrapper/pull/147)

ENDRELEASENOTES

- [x] Generalize map lookup functionality to work with things that look like a map (i.e. `map<K, V>`, `unordered_map<K, V>`, `vector<tuple<K, V>>`, ...
  - [x] Use this functionality in existing conversions
- [x] Implement map insertion functionality to work for things that look like a map
  - [x] Use this functionality in existing conversions 
- [x] Make the conversion functions templated on the map type
  - [x] Make conversion a header library 
- [x] Needs #27 
- [x] Needs #28 
- [x] Needs #29

### CODE REVIEW

**Comment by tmadlener** [comment]
@andresailer @jmcarcell This is in principle ready for review. The diff looks huge, but most of that is https://github.com/key4hep/k4EDM4hep2LcioConv/pull/21/commits/d5ae84607df9e614d7e7a2de731110aad3122f9a and https://github.com/key4hep/k4EDM4hep2LcioConv/pull/21/commits/e916827d72d75fc5e39c67c051eaf8bf5a351eb0 that shuffle quite a bit of code around. 

I keep it as WIP for now until I have also implemented the corresponding changes on the k4MarlinWrapper side, but I don't expect many changes here any longer.

**Comment by tmadlener** [comment]
The changes in the names of the maps in https://github.com/key4hep/k4EDM4hep2LcioConv/pull/21/commits/49dccf20f873d348e6d544929e722a6e8dcbaea6 makes this a breaking change for k4MarlinWrapper

I am not sure if it is worth to try and make this a slightly smoother transition, or whether we just bit the bullet.

File: k4EDM4hep2LcioConv/include/k4EDM4hep2LcioConv/MappingUtils.h [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+#define K4EDM4HEP2LCIOCONV_MAPPINGUTILS_H
+
+#include <optional>
+#include <algorithm>
+#include <vector>
+#include <tuple>
+#include <unordered_map>
+#include <type_traits>
+
+#if __has_include("experimental/type_traits.h")
```
Comment (andresailer) [inline k4EDM4hep2LcioConv/include/k4EDM4hep2LcioConv/MappingUtils.h:11]:
```suggestion
#if __has_include("experimental/type_traits")
```
But why do you have `#include <type_traits>`, and then again `#include <experimental/type_traits>`?

Comment (andresailer) [reply k4EDM4hep2LcioConv/include/k4EDM4hep2LcioConv/MappingUtils.h:11]: Ok, experimental/type_traits still have type_traits that are experimental.

**Pull request merged**